contentcuration/contentcuration/utils/recommendations.py

```diff
@@ -61,67 +68,200 @@ def create_backend(self) -> Backend:
+        """
+        Generates recommendations for the given request.
 
+        :param request: The request for which to generate embeddings.
+        :param cache: Whether to cache the embeddings request.
+        :return: The response containing the recommendations.
+        :rtype: EmbeddingsResponse
+        """
         if not self.backend.connect():
             raise errors.ConnectionError("Connection to the backend failed")
 
+        cached_response = self.response_exists(request)
+        if cached_response:
+            return cached_response
+
         try:
-            embed_topics_request = EmbedTopicsRequest(json=topics)
-            return self.backend.make_request(embed_topics_request)
+            response = self.backend.make_request(request)
         except Exception as e:
-            return EmbeddingsResponse(error=e)
+            logging.exception(e)
+            response = EmbeddingsResponse(error=e)
 
-    def embed_content(self, nodes: List[ContentNode]) -> EmbeddingsResponse:
+        if not response.error and cache:
+            self.cache_embeddings_request(request, response)
 
-        if not self.backend.connect():
-            raise errors.ConnectionError("Connection to the backend failed")
+        return response
 
+    def response_exists(self, request) -> Union[EmbeddingsResponse, None]:
+        """
+        Checks if a cached response exists for the given request.
+
+        :param request: The request for which to check if a cached response exists.
+        :return: The cached response if it exists, otherwise None.
+        :rtype: Union[EmbeddingsResponse, None]
+        """
         try:
-            resources = [self.extract_content(node) for node in nodes]
-            json = {
-                'resources': resources,
-                'metadata': {}
-            }
-            embed_content_request = EmbedContentRequest(json=json)
-            return self.backend.make_request(embed_content_request)
+            cache = list(RecommendationsCache.objects.filter(request=request).order_by('rank'))
+            data = [entry.response for entry in cache]
+            if len(data) > 0:
+                return EmbeddingsResponse(data=data)
+            else:
+                return None
         except Exception as e:
-            return EmbeddingsResponse(error=e)
+            logging.exception(e)
+            return None
+
+    def cache_embeddings_request(self, request: BackendRequest, response: BackendResponse) -> bool:
+        """
+        Caches the recommendations request and response.
+
+        :param request: The request to cache.
+        :param response: The response to cache.
+        :return: A boolean indicating whether the caching was successful.
+        :rtype: bool
+        """
+        try:
+            nodes = self._extract_data(response)
+            cache = [
+                RecommendationsCache(
+                    request=request,
+                    response=node,
```

Comment:
Caching a dict as `response` for pointing to the node doesn't give us full advantage of the cache. This means when we use the cache to fetch recommended nodes, we always have 2 queries. What can we do to make it only 1?

Reply:
A foreign key should suffice I think?

Reply:
Yes a FK to `ContentNode` in `kolibri_public`